src/rpc/core_rpc_server.cpp

```diff
@@ -2781,6 +2782,142 @@ namespace cryptonote
     return on_get_service_nodes(req_all, res, error_resp);
   }
   //------------------------------------------------------------------------------------------------------------------------------
+  /// Start with seed and perform a series of computation arriving at the answer
+  static uint64_t perform_blockchain_test_routine(const cryptonote::core& core,
+                                                  uint64_t max_height,
+                                                  uint64_t seed)
+  {
+    /// Should be sufficiently large to make it impractical
+    /// to query remote nodes
+    constexpr size_t NUM_ITERATIONS = 1000;
+
+    std::mt19937_64 mt(seed);
+
+    crypto::hash hash;
+
+    uint64_t height = 0;
+
+    for (auto i = 0u; i < NUM_ITERATIONS; ++i)
+    {
+      /// incorporate seed so that the results couldn't be cached;
+      /// increment seed to break cycles;
+      height = height ^ (seed++);
+      height = height % (max_height + 1);
+
+      hash = core.get_block_id_by_height(height);
+
+      using blob_t = cryptonote::blobdata;
+      using block_pair_t = std::pair<blob_t, block>;
+
+      /// pick a random byte from the block blob
+      std::vector<block_pair_t> blocks;
+      std::vector<blob_t> txs;
+      if (!core.get_blockchain_storage().get_blocks(height, 1, blocks, txs)) {
+        MERROR("Could not query block at requested height: " << height);
+        return 0;
+      }
+      const blob_t &blob = blocks.at(0).first;
+      const uint64_t byte_idx = service_nodes::uniform_distribution_portable(mt, blob.size());
+      uint8_t byte = blob[byte_idx];
+
+      /// pick a random byte from a random transaction blob if found
+      if (!txs.empty()) {
+        const uint64_t tx_idx = service_nodes::uniform_distribution_portable(mt, txs.size());
+        const blob_t &tx_blob = txs[tx_idx];
+
+        /// not sure if this can be empty, so check to be safe
+        if (!tx_blob.empty()) {
+          const uint64_t byte_idx = service_nodes::uniform_distribution_portable(mt, tx_blob.size());
+          const uint8_t tx_byte = tx_blob[byte_idx];
+          byte ^= tx_byte;
+        }
+
+      }
+
+      {
+        /// reduce hash down to 8 bytes
+        const uint64_t n1 = *reinterpret_cast<uint64_t*>(hash.data);
+        const uint64_t n2 = *reinterpret_cast<uint64_t*>(hash.data + 8);
+        const uint64_t n3 = *reinterpret_cast<uint64_t*>(hash.data + 16);
+        const uint64_t n4 = *reinterpret_cast<uint64_t*>(hash.data + 24);
```

Comment:
I get the endianess, but could you explain the alignment issues? Do you mean address of each character in the array could be aligned to the WORD size of the architecture in question, meaning there's potential padding inbetween each element?

For the endian swap, in this instance I would check the implementation of that boost endian swap, because the endian swap macros in the codebase are actually implemented in a sane way

Reply:
It's not a padding issue, it's that ARM (for example) won't allow reading a uint64_t that doesn't start on an 8-byte boundary in memory.  (x86 allows this, though takes a performance hit).  So basically if `hash.data` (and thus `hash.data + 8`) isn't at a 8-byte boundary ARM will fail to read the uint64_t value.  Whereas the memcpy doesn't have any such issue, and the 4-element array is guaranteed to meet the uint64_t alignment requirements.

> For the endian swap, in this instance I would check the implementation of that boost endian swap, because the endian swap macros in the codebase are actually implemented in a sane way

It's pretty sane: if you're on a little-endian architecture (which will almost always be the case) the call in a no-op; otherwise it's a call to the implementation-specific built-in (`__builtin_bswap64` under gcc/clang; `_byteswap_uint64` under MSVC; otherwise if you're on something more obscure it has a custom implementation that looks likely to be quite efficient).

Reply:
Good point about the endianness, will update. But I also don't understand the problem with alignment.

Reply:
Updated to respect endianness.